server/chat-commands/moderation.ts

```diff
@@ -95,70 +174,27 @@ export const commands: ChatCommands = {
 		const nextGroupName = nextGroup.name || "regular user";
 
 		for (const toPromote of users) {
-			const targetUser = Users.getExact(toPromote);
 			const userid = toID(toPromote);
-			const name = targetUser ? targetUser.name : this.filter(toPromote);
-			if (!name) continue;
-
 			if (!userid) return this.parse('/help roompromote');
-			if (userid.length > 18) {
-				this.errorReply(`User '${name}' does not exist (the username is too long).`);
-				continue;
-			}
-			if (!targetUser && !Users.isUsernameKnown(userid) && !force) {
-				this.errorReply(`User '${name}' is offline and unrecognized, and so can't be promoted.`);
-				continue;
-			}
-			if (targetUser && !targetUser.registered) {
-				this.errorReply(`User '${name}' is unregistered, and so can't be promoted.`);
-				continue;
-			}
 
-			let currentSymbol: GroupSymbol | 'whitelist' = room.auth.getDirect(userid);
-			if (room.auth.has(userid) && currentSymbol === Users.Auth.defaultSymbol()) {
-				currentSymbol = 'whitelist';
-			}
-			const currentGroup = Users.Auth.getGroup(currentSymbol);
-			const currentGroupName = currentGroup.name || "regular user";
-
-			if (currentSymbol === nextSymbol) {
-				this.errorReply(`User '${name}' is already a ${nextGroupName} in this room.`);
-				continue;
-			}
-			if (!user.can('makeroom')) {
-				if (currentGroup.id && !user.can(`room${currentGroup.id || 'voice'}` as 'roomvoice', null, room)) {
-					this.errorReply(`/${cmd} - Access denied for promoting/demoting ${name} from ${currentGroupName}.`);
-					continue;
-				}
-				if (nextSymbol !== ' ' && !user.can(`room${nextGroup.id || 'voice'}` as 'roomvoice', null, room)) {
-					this.errorReply(`/${cmd} - Access denied for promoting/demoting ${name} to ${nextGroupName}.`);
-					continue;
-				}
-			}
-			if (targetUser?.locked && room.persist && room.settings.isPrivate !== true && nextGroup.rank > 2) {
-				this.errorReply(`${name} is locked and can't be promoted.`);
+			const oldSymbol = room.auth.getDirect(userid);
+			let shouldPopup;
+			try {
+				shouldPopup = runPromote(user, room, userid, nextSymbol, toPromote, force);
+			} catch (err) {
+				if (err.name?.endsWith('ErrorMessage')) this.errorReply(err.message);
```

Comment:
shouldnt you still `throw new Chat.ErrorMessage(err.message);`?

Reply:
No, if you're promoting many users, you don't want one failure to stop the other promotions.